borg/repository.py

```diff
@@ -188,10 +191,23 @@ def commit(self, save_space=False):
         self.write_index()
         self.rollback()
 
-    def open_index(self, transaction_id):
+    def open_index(self, transaction_id, auto_recover=True):
         if transaction_id is None:
             return NSIndex()
-        return NSIndex.read((os.path.join(self.path, 'index.%d') % transaction_id).encode('utf-8'))
+        index_path = (os.path.join(self.path, 'index.%d') % transaction_id).encode('utf-8')
+        try:
+            return NSIndex.read(index_path)
+        except RuntimeError as re:
+            assert str(re) == 'hashindex_read failed'  # everything else means we're in *deep* trouble
```

Comment:
Good catch. In that case the later unlink() will explode, but that isn't correctly converted to InternalOSError yet, so would be silenced in an upper layer; not good.